### PR TITLE
[data]Add stratify parameter to train_test_split method

### PR DESCRIPTION
**Pull request by krishnakalyan3**
Add stratify parameter to train_test_split method 

## Related issue number
https://github.com/ray-project/ray/issues/53297

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by krishnakalyan3** [comment]
I was wondering If we need to implement a shuffle after the stratification?.

```
if shuffle:
    train_ds = train_ds.random_shuffle(seed=seed)
    test_ds = test_ds.random_shuffle(seed=seed)
```

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -2311,6 +2321,52 @@ def train_test_split(
...
+        Args:
+            ds: The dataset to split.
+            test_size: Test size as int or float.
+            stratify: Column name to use for stratified sampling.
+            
+        Returns:
+            Train and test subsets as two MaterializedDatasets.
+        """
+        # Normalize test_size to float (only materialize if needed)
+        if isinstance(test_size, int):
```
Comment (raulchen) [inline python/ray/data/dataset.py:2339]: can you unify the validation code with the original train_test_split to avoid duplication?

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -2311,6 +2321,52 @@ def train_test_split(
...
+            )
+        
+        def add_train_flag(group_batch):
+            import pyarrow as pa
+            # group_batch is already a pandas DataFrame
+            df = group_batch
+            n = len(df)
+            test_count = int(n * test_size)
+            df["is_train"] = [True] * (n - test_count) + [False] * test_count
+            return pa.Table.from_pandas(df)
```
Comment (raulchen) [inline python/ray/data/dataset.py]: I think the default batch format should be numpy, not pandas
And we don't need to manually convert the batch to a pyarrow table. This UDF is supposed to return a batch.
you can just do this:
```
            n = len(group_batch)
            test_count = int(n * test_size)
            group_batch["is_train"] = [True] * (n - test_count) + [False] * test_count
             return group_batch
```

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -2311,6 +2321,52 @@ def train_test_split(
...
+        
+        split_ds = ds.groupby(stratify).map_groups(add_train_flag)
+        
+        train_ds = split_ds.filter(lambda row: row["is_train"]).drop_columns(["is_train"])
```
Comment (raulchen) [inline python/ray/data/dataset.py]: use a more special column name to avoid collision with user columns. 
And please define a constant

File: python/ray/data/dataset.py [hunk 4]
```diff
@@ -2311,6 +2321,52 @@ def train_test_split(
...
+        def add_train_flag(group_batch):
+            import pyarrow as pa
+            # group_batch is already a pandas DataFrame
+            df = group_batch
+            n = len(df)
+            test_count = int(n * test_size)
+            df["is_train"] = [True] * (n - test_count) + [False] * test_count
+            return pa.Table.from_pandas(df)
+        
+        split_ds = ds.groupby(stratify).map_groups(add_train_flag)
```
Comment (raulchen) [inline python/ray/data/dataset.py]: We can materialize this dataset to avoid duplicated computations.

File: python/ray/data/tests/test_split.py [hunk 5]
```diff
@@ -776,6 +776,37 @@ def test_train_test_split(ray_start_regular_shared):
...
+    # Check that class proportions are preserved
+    train_labels = [row["label"] for row in train.take()]
+    test_labels = [row["label"] for row in test.take()]
+    
+    train_label_counts = {label: train_labels.count(label) for label in ["A", "B", "C"]}
+    test_label_counts = {label: test_labels.count(label) for label in ["A", "B", "C"]}
+    
+    # Each class should have exactly 1 sample in each split
+    assert train_label_counts == {"A": 1, "B": 1, "C": 1}
+    assert test_label_counts == {"A": 1, "B": 1, "C": 1}
```
[the remark this answers is not in the document]
Comment (raulchen) [reply python/ray/data/tests/test_split.py]: yes, please also test imbalanced distributions and edge cases.

File: python/ray/data/dataset.py [hunk 6]
```diff
@@ -2294,6 +2298,12 @@ def train_test_split(
 
         if not isinstance(test_size, (int, float)):
             raise TypeError(f"`test_size` must be int or float got {type(test_size)}.")
+        
+        # Handle stratified splitting
+        if stratify is not None:
+            return self._stratified_split(ds, test_size, stratify)
```
Comment (raulchen) [inline python/ray/data/dataset.py]: I think we should  disallow specifying both shuffle and stratify at the same time.

File: python/ray/data/dataset.py [hunk 7]
```diff
@@ -2294,6 +2301,47 @@ def train_test_split(
...
+                        f"than the size of the dataset ({ds_length}). "
+                        f"Got {test_size}."
+                    )
+                test_size = test_size / ds_length
+            
+            if test_size <= 0 or test_size >= 1:
+                raise ValueError(
+                    "For stratified splitting, test_size must be between 0 and 1 "
+                    f"(or equivalent int). Got {test_size}."
+                )
```
Comment (raulchen) [inline python/ray/data/dataset.py]: oh, I meant unifying the `test_size` validation code. it's duplicated with the following
```
        if isinstance(test_size, float):
            if test_size <= 0 or test_size >= 1:
                raise ValueError(
                    "If `test_size` is a float, it must be bigger than 0 and smaller "
                    f"than 1. Got {test_size}."
                )
            return ds.split_proportionately([1 - test_size])
        else:
            ds_length = ds.count()
            if test_size <= 0 or test_size >= ds_length:
                raise ValueError(
                    "If `test_size` is an int, it must be bigger than 0 and smaller "
                    f"than the size of the dataset ({ds_length}). "
                    f"Got {test_size}."
                )
            return ds.split_at_indices([ds_length - test_size])
```

Putting the stratified split in a separate function is good.

**Review by raulchen — COMMENTED** [review]
LGTM except one small comment

File: python/ray/data/dataset.py (hunk 7 above)
Comment (krishnakalyan3) [reply python/ray/data/dataset.py]: ah got it. thanks

Comment (krishnakalyan3) [reply python/ray/data/dataset.py]: @raulchen can you please have a look if this is better?.

**Review by raulchen — APPROVED** [review]
LGTM. please fix the lint and DCO errors. then we can merge it

**Comment by gvspraveen** [comment]
@krishnakalyan3 Thanks for your contribution. This feature is useful.

Can you fix DCO and lint when you get a chance?

File: python/ray/data/dataset.py [hunk 8]
```diff
@@ -2256,6 +2259,7 @@ def train_test_split(
         *,
         shuffle: bool = False,
         seed: Optional[int] = None,
+        stratify: Optional[str] = None,
```
Comment (abrarsheikh) [inline python/ray/data/dataset.py:2262]: consider supporting list of cols similar to what sklearn supports https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

Comment (krishnakalyan3) [reply python/ray/data/dataset.py:2262]: Hi @abrarsheikh ,
By default we stratify based on the target or the label column name. This is different for different datasets and dynamic. I am not sure what you exactly mean. Could you please clarify?.

Comment (raulchen) [reply python/ray/data/dataset.py:2262]: I think they want to stratify on multiple columns. 
That makes sense, and we can do that in a follow-up PR.

Comment (abrarsheikh) [reply python/ray/data/dataset.py:2262]: not sure how successfully we can do that as a followup, because that is a breaking change to the API, from `str` -> List[str]. Just want to point that out.
